### DISCUSSION TITLE
How to scroll datagrid a page at a time from code-behind?

### DISCUSSION THREAD

**Discussion by mkj-stonehill** [AvaloniaUI/Avalonia#11237] [Q&A]
My application has a screen showing results in a DataGrid.  It's running on Linux, using a small (7") touchscreen; since it's extremely difficult to manipulate the scroll bar on the DataGrid using the touchscreen, I wanted to add a couple of buttons to scroll the results up and down a screen full at a time.  But I cannot see any way to do it.

The VerticalScrollBar on the DataGrid is not public, so I can't see any way to access it from code-behind.

The DataGrid doesn't provide any way to tell how many rows are visible at a time, so I would have to hard-code a value and use ScrollIntoView.

But ScrollIntoView doesn't really work in this case, since it doesn't scroll the specified item to the top of the screen;  e.g. if the DataGrid is showing 10 items, and I call ScrollIntoView with the 11th item, it just scrolls one line (to put the 11th item at the bottom of the screen).

Is there any straight-forward way to do this?  Or even a convoluted but reliable way?

Thanks in advance for any suggestions,

--mkj

**Comment by timunie** [comment]
@mkj-stonehill in Avalonia, you can alsways explore the VisualTree to access any sub control of a control, in your case the vertical scroll viewer. Use DevTools [F12] to see the VisualTree rendered. 

Visual helpers: https://reference.avaloniaui.net/api/Avalonia.VisualTree/VisualExtensions/

**Comment by mkj-stonehill** [comment]
This was really helpful; thank you!  I had initially tried FindControl<ScrollBar>("PART_VerticalScrollbar"), but it always returned null (maybe because ScrollBar is not a Control, but a Primitive?).  Anyway, FindDescendantOfType<ScrollBar> found it.

However, the ScrollBar doesn't seem to have function I can call to trigger a scroll.  And, even if it did, looking at the source code for DataGrid, it only handles small increment/decrement (which scrolls one row at a time), not large increment/decrement.  So it looks like I am just out of luck, unless I want to modify the source.

I have seen some postings that say "avoid DataGrid at all costs; use TreeDataGrid instead".  Any thoughts on whether that is worth pursuing?  Or am I going to right back in the same boat?

**Comment by wgraaf** [comment]
I have a similar issue: 

I need to capture and restore scroll positions of a datagrid on a tab control.
By default, the scroll position gets reset upon switching tabs.
